api/types/webauthn/webauthn.proto

```diff
@@ -0,0 +1,257 @@
+syntax = "proto3";
+
+// Package WebAuthn maps WebAuthn messages to protocol buffers.
+//
+// The mapping is designed to match both the WebAuthn specification and the
+// capabilities of current browser implementations.
+//
+// REST-based Teleport APIs will make an effort to transmit or embed JSON
+// messages matching the github.com/duo-labs/webauthn reference implementation,
+// to allow for easy browser integration. gRPC APIs are not meant for REST use
+// and thus make no such promises, although the correspondence should be
+// obvious.
+//
+// Note that, ordinarily, various fields in WebAuthn messages are encoded using
+// "RawURLEncoding" (aka, base64 URL encoding without padding). This is not the
+// case for _any_ of the fields mapped here, all bytes fields are transmitted
+// raw/unencoded.
+package webauthn;
+
+import "gogoproto/gogo.proto";
+
+option (gogoproto.marshaler_all) = true;
+option (gogoproto.unmarshaler_all) = true;
+
+// -----------------------------------------------------------------------------
+// WebAuthn messages used by server storage.
+// -----------------------------------------------------------------------------
+
+// SessionData stored by the Relying Party during authentication ceremonies.
+// Mirrors https://pkg.go.dev/github.com/duo-labs/webauthn/webauthn#SessionData.
+message SessionData {
+    // Raw challenge used for the ceremony.
+    bytes challenge = 1 [ (gogoproto.jsontag) = "challenge,omitempty" ];
+    // Raw User ID.
+    bytes user_id = 2 [ (gogoproto.jsontag) = "userId,omitempty" ];
+    // Raw Credential IDs of the credentials allowed for the ceremony.
+    repeated bytes allow_credentials = 3 [ (gogoproto.jsontag) = "allowCredentials,omitempty" ];
+    // Requested user verification.
+    // https://www.w3.org/TR/webauthn-2/#enum-userVerificationRequirement.
+    string user_verification = 4 [ (gogoproto.jsontag) = "userVerification,omitempty" ];
```

Comment:
do we need to transmit/store this?
if we always set `preferred` internally, there's no reason to expose it in the API

same below, are all of those fields needed?
can some of them be inferred by the proxy (like timeouts, rp_id, extensions, etc)?

with protobufs, we can always add more fields later as needed.
ideally, the API surface should be as small as possible to reduce attack surface.

Reply:
I'll do another pass, give me a few minutes.

For the ones you mentioned:

* Removed user_verification from SessionData (fair enough)
* Timeout is a hint to clients, I think it's best to keep it as we may change it
* I'd rather not have clients guessing RPIDs either, seems like a recipe for trouble
* Extensions we need to pass on - specifically, the appid extension

Reply:
I've removed a few fields I think we can get away without, PTAL.